python/ray/train/data_parallel_trainer.py

```diff
@@ -520,16 +521,30 @@ def clear_lazy_checkpoint_marker():
         checkpoint_strategy.num_to_keep = None
         checkpoint_strategy.checkpoint_score_attribute = None
 
+        initial_checkpoint = self._get_initial_checkpoint()
+
+        latest_checkpoint_index = 0
+        if _use_storage_context() and self._checkpoint_for_restoration:
+            checkpoint_dir_name = os.path.basename(
+                self._checkpoint_for_restoration.path
+            )
+            # Need to add 1 because latest_checkpoint_index is the index of the
+            # next checkpoint to be saved.
+            latest_checkpoint_index = (
+                StorageContext._parse_checkpoint_index(checkpoint_dir_name) + 1
+            )
```

Comment:
Only restore the "checkpoint index" when `_checkpoint_for_restoration` is set (aka cases 1 and 2).

Completely fresh run (keeping no state) if starting from `resume_from_checkpoint`.

Maybe too hacky here.

Reply:
This might be more readable:
1. Always use resume_from_checkpoint
2. Have a separate Optional[int] initial_checkpoint_index that can be populated by the trial in the case of autoresume.

Reply:
There's now no more `initial_checkpoint_index` that gets passed around. It just comes from the driver's `trial.storage.current_checkpoint_index` now.

I did keep the `_checkpoint_for_restoration` vs. `resume_from_checkpoint` distinction though, since I thought overwriting that user-specified attribute on restore was confusing. Happy to revert this back though.

Reply:
Looks much better.

> I did keep the _checkpoint_for_restoration vs. resume_from_checkpoint distinction though, since I thought overwriting that user-specified attribute on restore was confusing. Happy to revert this back though.

I do feel like this is still strange though, since from the user code POV there is no real difference between these two fields--- they are both defining which checkpoint should be resumed from. Perhaps we can just rename the class argument to `starting_checkpoint` or something like this to clarify that it can be provided for different reasons?

Reply:
Yep, can just change the single attribute name to be clearer.